EOS-13787:[Enabler] NFS ADDB: Explore and implement add2db.py for NFS

pratyush-seagate opened the pull request:
# EOS-13787:[Enabler] NFS ADDB: Explore and implement add2db.py for NFS

## Checklist
- [x] **Compilation:** _This patch does not break compilation_
- [x] **Merge conflicts:** _This patch has been squashed and re-based, it can be merged using fast-forward merge_
- [] **Code review:** _All discussions have been resolved_
- [x] **Sanity Testing:** _All Unit tests are passing and able to do mount and io operations works from NFS client_
- [x] **Documentation:** _This patch and merge request have up to date description_
- [x] **Unit Testing and debugging:** _Both single and multi node changes are unit tested_

## Commit Message
        commit 8a96b6ef51123e771030e9c7cd8acf80c65998e0
        Author: pratyush-seagate <pratyush.k.khan@seagate.com>
        Date:   Sun Oct 4 15:40:30 2020 -0600

        EOS-13787:[Enabler] NFS ADDB: Explore and implement add2db.py for NFS
        List of modified files:
                new file:   addb2db_nfs.py
        Change description:
                Add python script to read and convert the text performance dump to sql database file for NFS
        Unit test:
                Based on EOS-8947, using the new script, read the filtered text dump file containing the new
                nfs-cortxfs performance counters (state, map and attribute) and create a sql3 database file.
                Verify the sql3 database file is ok.
                test O/P:
                        cd /home/751521/EOS-13787
                        m0addb2dump -p /usr/lib64/libcortx-utils.so -- /$(ls -t addb* | head -n 1 | sed 's/://')/o/100000000000000\:2 | grep fsuser &> perfc_raw_dump
                        python3 /home/751521/repos/github/repo4/cortx-posix/utils/c-utils/src/addb/addb2db_nfs.py --dumps ./perfc_raw_dump --db DB2
                        # ll
                        total 13664
                        -rw-r--r-- 1 root root 11476077 Oct  4 15:37 addb_27527.tar.gz
                        -rw-r--r-- 1 root root   963584 Oct  4 15:39 DB2
                        -rw-r--r-- 1 root root      395 Oct  4 15:39 logfile.txt
                        -rw-r--r-- 1 root root  1541590 Oct  4 15:38 perfc_raw_dump

                        sqlite3 ./DB2
                        sqlite> .schema
                                CREATE TABLE "entity_attributes" ("id" INTEGER NOT NULL PRIMARY KEY, "pid" INTEGER NOT NULL, "time" INTEGER NOT NULL, "tsdb_mod" TEXT NOT NULL, "fn
                                CREATE TABLE "entity_maps" ("id" INTEGER NOT NULL PRIMARY KEY, "pid" INTEGER NOT NULL, "time" INTEGER NOT NULL, "tsdb_mod" TEXT NOT NULL, "fn_tag"
                                CREATE TABLE "entity_states" ("id" INTEGER NOT NULL PRIMARY KEY, "pid" INTEGER NOT NULL, "time" INTEGER NOT NULL, "tsdb_mod" TEXT NOT NULL, "fn_tag

                        sqlite> select * from "entity_states";
                        ....
                        5330|14|1601437968920457574|fsuser_state|fsal_read|state|30133|finish

                        sqlite> select * from "entity_attributes";
                        ....
                        6273|14|1601437968920456127|fsuser_attribute|fsal_read|attribute|30133|read_result_major_code|0

                        sqlite> select * from "entity_maps";
                        ....
                        2510|14|1601437968918784740|fsuser_map|kvs_alloc|map|16|0|30133

                Note: python3 libs needed: sudo pip3 install numpy plumbum  tqdm peewee

vimalraghwani20 reviewed (commented):
IMO, We need to follow the folder structure of S3 for decoder plugin and scripts.
addb/plugin/
addb/addb-py/chronometry/

vimalraghwani20 reviewed (approved):
I tried using this script and able to create a database with the records.
Patch looks good to me

Shreya-18 commented on c-utils/src/addb/addb2db_nfs.py:
```diff
@@ -0,0 +1,316 @@
+
+class ADDB2PPNFS:
+    @staticmethod
+    def clean_yaml(yml):
```
I see this `clean_yaml(yml):` is not being used anywhere. Could you please check whether it can be removed?

Shreya-18 commented on c-utils/src/addb/addb2db_nfs.py:
```diff
@@ -0,0 +1,316 @@
+
+import argparse
+import logging
+import yaml
```
Also, can this import yaml can be removed if not used?

Shreya-18 commented on c-utils/src/addb/addb2db_nfs.py:
```diff
@@ -0,0 +1,316 @@
+# ======================================================================
+
+class profiler:
+    def __init__(self, what):
```
Can we have a one-liner comment describing the **user-defined** context manager class here?

pratyush-seagate commented:
@Shreya-18  addressed all the review comments.
@sachinpunadikar  @AtitaShirwaikar  all the codacy comments are resolved except the last one, this is due to https://github.com/PyCQA/pydocstyle/issues/141
IMO, we should disable this check due to the conflicting rules.

![image](https://user-images.githubusercontent.com/67319330/95184455-03f19000-07e5-11eb-95ff-605770040cef.png)

Shreya-18 reviewed (approved):
Looks good to me. Nice usage of context managers for resource handling. +1 from my side.

sachinpunadikar commented on c-utils/src/addb/addb2db_nfs.py:
```diff
@@ -0,0 +1,326 @@
+#
```
wondering whether should be part of py-utils folder ?
or check with S3 team on how they are processing ADDB captured data

pratyush-seagate replied:
Thanks Sachin, let me refactor this code and update the PR.

pratyush-seagate replied:
Hi Sachin,
The python script is now moved to py-utils and added to py-utils build.
Regarding the refactoring, after looking into details, it might take more time to decouple the NFS performance counters and create sub-modules. The S3 team's scripts are also not refactored. If it is ok with you, can I create a JIRA story to track the future refactoring work to make these scripts more modular?

pratyush-seagate replied:
The JIRA story created to track the end to end documentation work: https://jts.seagate.com/browse/EOS-14033
The JIRA story created to track the refactoring work for python scripts: https://jts.seagate.com/browse/EOS-14034

sachinpunadikar commented on py-utils/src/utils/addb/addb2db_nfs.py:
```diff
@@ -0,0 +1,324 @@
+#
```
Shouldn't the name of script be adddb2db_cortxfs ? I do not see anywhere it is only for NFS

pratyush-seagate replied:
Done

sachinpunadikar commented on py-utils/setup.py:
```diff
@@ -39,6 +39,7 @@
       description='Common Python utilities for CORTX',
       package_dir={'cortx': 'src'},
       packages=['cortx', 'cortx.utils',
+                'cortx.utils.addb',
```
Remove from here .. as this will affect the package cortx.utils , which is being used by other teams. We need to be in agreement with other team before changing anything there.
I hope we can use the scripts in standalone method

pratyush-seagate replied:
Done

sachinpunadikar reviewed (changes requested):
Please work on review comments

pratyush-seagate commented:
> 
> 
> Please work on review comments

Done

sachinpunadikar reviewed (approved):
+1
